Flaky API Test: TestWTF.WTF_SynchronizedFixedQueue.ProduceOnly

fujii opened the pull request:
#### 5b9202cb4147f34728e00b156392ac54a0b9a0dc
<pre>
Flaky API Test: TestWTF.WTF_SynchronizedFixedQueue.ProduceOnly
<a href="https://bugs.webkit.org/show_bug.cgi?id=195161">https://bugs.webkit.org/show_bug.cgi?id=195161</a>

Reviewed by Said Abou-Hallawa.

TestWTF.WTF_SynchronizedFixedQueue.ProduceOnly was randomly timing out
on Windows. Produce thread was blocked by &quot;m_upperQueue-&gt;enqueue(...)&quot;
because m_upperQueue was full. This happened if the Produce thread was
running quicker than the main thread.

Doubled BufferSize of ToUpperConverter not to block the Produce thread
in the ProduceOnly test. And, run the test until produceCount reaches
the item count.

192736@main (bug#176056) was the previous attempt to fix this dead
lock by slowing down the Produce thread. This change is no longer
needed. Reverted it.

* Tools/TestWebKitAPI/Tests/WTF/SynchronizedFixedQueue.cpp:
(TestWebKitAPI::ToUpperConverter::startProducing):
(TestWebKitAPI::ToUpperConverter::stopProducing):

Canonical link: <a href="https://commits.webkit.org/263664@main">https://commits.webkit.org/263664@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cfbac4d59fd2c714f766f22f2bb32836601f597a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5371 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5508 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5697 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6909 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5407 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5371 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5735 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5487 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/6215 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5470 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5540 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4805 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6937 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3019 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4811 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/11967 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4885 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4890 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6513 "258 api tests failed or timed out") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5324 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4370 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4788 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1288 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8885 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5149 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Tools/TestWebKitAPI/Tests/WTF/SynchronizedFixedQueue.cpp:
```diff
@@ -123,6 +118,7 @@ class ToUpperConverter {
             return;
 
         m_lowerQueue->close();
+        m_upperQueue->close();
```
Yes this will fix the issue but I think it is a little bit ugly. As you explained in the ChangeLog, the ProduceOnly problem happens because the main thread is blocked on
```
    m_produceCloseSemaphore.wait();
```
while the produceQueue() is blocked on 
```
    m_upperQueue->enqueue(toUpper(lower));
```
`m_upperQueue` is full since the `consumeQueue` is not running. Your solution fixes this issue by closing `m_upperQueue` so, `produceQueue()` can signal `m_produceCloseSemaphore`.

Alternatively I suggest the following:
1. Increase the size of `ToUpperConverter` in ProduceOnly to 8U. So no waiting in `produceQueue()`.
2. Change the type of `m_produceCount` and `m_consumeCount` to be `std::atomic<size_t>`
3. Change the return type of produceCount() and consumeCount() to be `const std::atomic<size_t>&`
4. Spin in `ProduceOnly` till `produceCount()` is equal to `count`.
5. Please Add these checks at the end of `ProduceOnly`
```
    EXPECT_EQ(converter.produceCount(), count);
    EXPECT_EQ(converter.consumeCount(), 0U);
```

Also in the `Limits` sub-test, please replace the `sleep_for()` with this spin:

```
    while (converter.consumeCount() < count)
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
```

fujii replied:
Fixed. Thank you very much.

fujii committed:
Flaky API Test: TestWTF.WTF_SynchronizedFixedQueue.ProduceOnly
https://bugs.webkit.org/show_bug.cgi?id=195161

Reviewed by Said Abou-Hallawa.

TestWTF.WTF_SynchronizedFixedQueue.ProduceOnly was randomly timing out
on Windows. Produce thread was blocked by "m_upperQueue->enqueue(...)"
because m_upperQueue was full. This happened if the Produce thread was
running quicker than the main thread.

Doubled BufferSize of ToUpperConverter not to block the Produce thread
in the ProduceOnly test. And, run the test until produceCount reaches
the item count.

192736@main (bug#176056) was the previous attempt to fix this dead
lock by slowing down the Produce thread. This change is no longer
needed. Reverted it.

* Tools/TestWebKitAPI/Tests/WTF/SynchronizedFixedQueue.cpp:
(TestWebKitAPI::ToUpperConverter::startProducing):
(TestWebKitAPI::ToUpperConverter::stopProducing):

Canonical link: https://commits.webkit.org/263664@main
